ql/src/test/org/apache/hadoop/hive/ql/exec/TestGetPartitionInBatches.java

```diff
@@ -0,0 +1,245 @@
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testSmallNumberOfPartitions() throws Exception {
+        HiveMetaStoreClient spyMSC = spy(db);
+        hive.setMSC(spyMSC);
+        hive.getAllPartitionsOf(hive.getTable(dbName, tableName),100, 2, 0);
+        ArgumentCaptor<GetPartitionsByNamesRequest> req = ArgumentCaptor.forClass(GetPartitionsByNamesRequest.class);
+        // there should be 1 call to get partitions
+        verify(spyMSC, times(1)).getPartitionsByNames(req.capture());
+        Assert.assertEquals(30, req.getValue().getNames().size());
+    }
+
+    /**
+     * Tests the retries exhausted case when getAllPartitionsOf method call always keep throwing
+     * HiveException. The batch sizes should exponentially decreased based on the decaying factor and
+     * ultimately give up when it reaches 0
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testRetriesExhaustedBatchSize() throws Exception {
+        HiveMetaStoreClient spyMSC = spy(db);
+        hive.setMSC(spyMSC);
+        doThrow(MetaException.class).when(spyMSC).getPartitionsByNames(any());
+        try {
+            hive.getAllPartitionsOf(hive.getTable(dbName, tableName), 30, 2, 0);
+        } catch (Exception ignored) {}
+        ArgumentCaptor<GetPartitionsByNamesRequest> req = ArgumentCaptor.forClass(GetPartitionsByNamesRequest.class);
+        // there should be 5 call to get partitions with batch sizes as 30, 15, 7, 3, 1
+        verify(spyMSC, times(5)).getPartitionsByNames(req.capture());
+        List<GetPartitionsByNamesRequest> apds = req.getAllValues();
+        Assert.assertEquals(5, apds.size());
+
+        Assert.assertEquals(30, apds.get(0).getNames().size());
+        Assert.assertEquals(15, apds.get(1).getNames().size());
+        Assert.assertEquals(7, apds.get(2).getNames().size());
+        Assert.assertEquals(3, apds.get(3).getNames().size());
+        Assert.assertEquals(1, apds.get(4).getNames().size());
+    }
+
+    /**
+     * Tests the maximum retry attempts provided by configuration
+     * @throws Exception
+     */
+    @Test
+    public void testMaxRetriesReached() throws Exception {
+        HiveMetaStoreClient spyMSC = spy(db);
+        hive.setMSC(spyMSC);
+        doThrow(MetaException.class).when(spyMSC).getPartitionsByNames(any());
+        try {
+            hive.getAllPartitionsOf(hive.getTable(dbName, tableName), 30, 2, 2);
+        } catch (Exception ignored) {}
+        ArgumentCaptor<GetPartitionsByNamesRequest> req = ArgumentCaptor.forClass(GetPartitionsByNamesRequest.class);
+        // there should be 2 call to get partitions with batch sizes as 30, 15
+        verify(spyMSC, times(2)).getPartitionsByNames(req.capture());
+        List<GetPartitionsByNamesRequest> apds = req.getAllValues();
+        Assert.assertEquals(2, apds.size());
+
+        Assert.assertEquals(30, apds.get(0).getNames().size());
+        Assert.assertEquals(15, apds.get(1).getNames().size());
+    }
+}
```

Comment:
Add test that it is indeed getting you all the partitions(and no duplicate) after failure to a batch fetch in between and hence reducing the batch size/2 during retry.

Reply:
Added a test for same as well.